contracts/utils/math/Math.sol

```diff
@@ -339,38 +339,113 @@ library Math {
+     * This method is based on Newton's method for computing square roots; the algorithm is restricted to only
+     * using integer operations.
      */
     function sqrt(uint256 a) internal pure returns (uint256) {
-        if (a == 0) {
-            return 0;
-        }
-
-        // For our first guess, we get the biggest power of 2 which is smaller than the square root of the target.
-        //
-        // We know that the "msb" (most significant bit) of our target number `a` is a power of 2 such that we have
-        // `msb(a) <= a < 2*msb(a)`. This value can be written `msb(a)=2**k` with `k=log2(a)`.
-        //
-        // This can be rewritten `2**log2(a) <= a < 2**(log2(a) + 1)`
-        // → `sqrt(2**k) <= sqrt(a) < sqrt(2**(k+1))`
-        // → `2**(k/2) <= sqrt(a) < 2**((k+1)/2) <= 2**(k/2 + 1)`
-        //
-        // Consequently, `2**(log2(a) / 2)` is a good first approximation of `sqrt(a)` with at least 1 correct bit.
-        uint256 result = 1 << (log2(a) >> 1);
-
-        // At this point `result` is an estimation with one bit of precision. We know the true value is a uint128,
-        // since it is the square root of a uint256. Newton's method converges quadratically (precision doubles at
-        // every iteration). We thus need at most 7 iteration to turn our partial result with one bit of precision
-        // into the expected uint128 result.
         unchecked {
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            result = (result + a / result) >> 1;
-            return min(result, a / result);
+            // Take care of easy edge cases when a == 0 or a == 1
+            if (a <= 1) {
+                return a;
+            }
+
+            // In this function, we use Newton's method to get a root of `f(x) := x² - a`. This is also known as
+            // Heron's method. This involves building a sequence x_n that converges toward sqrt(a). For each
+            // iteration x_n, we define `ε_n = | x_n - sqrt(a) |`. This represents the error between the current value
+            // and the result.
+            //
+            // For our first estimation, we consider `e` the smallest power of 2 which is bigger than the square root
+            // of the target. (i.e. `2**(e-1) ≤ sqrt(a) < 2**e`). We know that `e ≤ 128` because `(2¹²⁸)² = 2²⁵⁶` is
+            // bigger than any uint256.
+            //
+            // By noticing that
+            // `2**(e-1) ≤ sqrt(a) < 2**e → (2**(e-1))² ≤ a < (2**e)² → 2**(2*e-2) ≤ a < 2**(2*e)`
+            // we can deduce that `e - 1` is `log2(a) / 2`. We can thus compute `x_n = 2**(e-1)` using a method similar
+            // to the msb function.
```

Comment:
If we reference Heron's method at the top, I'd remove this explanation as well, and just mention that the method will leave us an estimation that's less than or equal to the actual sqrt.

Reply:
> estimation that's less than or equal to the actual sqrt.

Which estimate? `2**(e-1)`? When we say its less than or equal to the sqrt, do we say by how much? If we don't say which one, and what its property are, then we might just say nothing. 

I think this is where we diverge. In my opinion, having vague comment is the worst. They raise question. They don't provide answers, and they end up being more confusing than helpfull.

Reply:
In my opinion, the code becomes clear as long as the comment indicates that it's making an estimation. I agree it's worth mentioning since it's a proof but my initial thought is that we don't need a formal one.

After our convo in Slack I agree the proof is valuable. I didn't want to bloat the implementation with comments but I thought we might allow this only in the Math.sol library.

Solving